### PR TITLE
Add CirrusCI for testing `patrol develop`

### PR DESCRIPTION
**Pull request by bartekpacia**

### CODE REVIEW

**Comment by bartekpacia** [comment]
Status: some tests fail for unknown reasons, and we don't have any screen recordings.

I think it's time to move our internal screen recorder shell script (from `customer:budowa`) to [mobile-tools](https://github.com/leancodepl/mobile-tools), and use it to record videos.

**Comment by bartekpacia** [comment]
We can use CirrusCI to test `patrol develop`, because CirrusCI machines are more powerful than GitHub Actions.

File: packages/patrol/example/patrol_develop_test.dart [hunk 1]
```diff
@@ -52,14 +52,24 @@ void main() async {
...
+        print('isReloaded: $isReloaded');
+        print('Running file:');
+        print(runningTestFile.readAsStringSync());
+        print('End of the running file');
+        print('Exiting with exit code 1');
+        exit(1);
+      });
+    } else {
+      inactivityTimer = Timer(Duration(minutes: 10), () {
+        print('Ten minutes of inactivity, something went wrong...');
```
Comment (bartekpacia) [inline packages/patrol/example/patrol_develop_test.dart:69]: Let's assign the duration to some variable, so there's no duplication

File: .github/workflows/label_pull_request.yaml [hunk 2]
```diff
@@ -12,6 +12,6 @@ jobs:
         uses: actions/checkout@v3
 
       - name: Label Pull Request
-        uses: actions/labeler@v4
```
Comment (bartekpacia) [inline .github/workflows/label_pull_request.yaml]: why?

File: .cirrus.yml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    - flutter precache
+  generate_gradlew_script: |
+    cd packages/patrol/example
+    flutter build apk --target lib/main.dart --debug --flavor=does-not-exist &
+    buildpid="$!"
+    while [ ! -e "android/gradlew" ]; do
+        echo "waiting for gradlew to generate..."
+        sleep 2
+    done
+    kill $buildpid
```
Comment (bartekpacia) [inline .cirrus.yml:27]: that's nice

File: packages/patrol/example/integration_test/example_test_fake.dart [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      await $.native.openApp();
+
+      expect($(#counterText).text, '1');
+      await $(FloatingActionButton).tap();
+
+      expect($(#counterText).text, '2');
+      expect($('Hello, fail here!'), findsOneWidget);
+    },
+  );
+}
```
Comment (bartekpacia) [inline packages/patrol/example/integration_test/example_test_fake.dart:27]: I don't want this file in `integration_test`.

Let's create it during the test:

```dart
const failingTest = '''

import 'package:flutter/material.dart';

import 'common.dart';

void main() {
  patrol(
    'This test is used to `test patrol develop`',
    ($) async {
      await createApp($);

      await $(FloatingActionButton).tap();
      expect($(#counterText).text, '1');

      await $(#textField).enterText('Hello, Flutter!');
      expect($('Hello, Flutter!'), findsOneWidget);

      await $.native.pressHome();
      await $.native.openApp();

      expect($(#counterText).text, '1');
      await $(FloatingActionButton).tap();

      expect($(#counterText).text, '2');
      expect($('Hello, fail here!'), findsOneWidget);
    },
  );
}
''';


```

**Comment by bartekpacia** [comment]
I'd prefer to have this test run on every PR, not every 12h.

File: .cirrus.yml [hunk 5]
```diff
@@ -78,4 +75,4 @@ test_ios_task:
   patrol_test_script: |
     cd packages/patrol/example
     flutter pub get
-    dart run patrol_develop_test.dart --device "$UDID"
+    dart run patrol_develop_test.dart
```
Comment (bartekpacia) [inline .cirrus.yml:78]: I'm pretty sure that `dart patrol_develop_test.dart` is shorter and does the same

File: .cirrus.yml [hunk 6]
```diff
@@ -91,37 +91,37 @@ test_windows_task:
...
+    echo "hw.lcd.width=1080" >> "$EMULATOR_CONFIG_PATH"
 
   start_emulator_background_script:
     - emulator @MyAVD -verbose -no-snapshot-save -no-window -noaudio -no-boot-anim -accel on
   wait_for_emulator_script:
     - android-wait-for-emulator
   patrol_develop_script: |
-    cd packages\\patrol\\example
+    cd packages\patrol\example
     dart run patrol_develop_test.dart
```
Comment (bartekpacia) [inline .cirrus.yml:131]: add trailing newline

**Comment by bartekpacia** [comment]
@fylyppo 

```yaml
container:
  image: cirrusci/flutter:stable
  cpu: 8
  memory: 32G
```

This is no good.  We need to shrink our resource usage as much as possible because that way we are more likely to get scheduled by CirrusCI. The more resources we request, the long we'll wait for a worker with that high resources.

File: dev/cli_tests/patrol_develop_test.dart [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    final stringOutput = output.toString();
+
+    if (isFirstTestPassed == false &&
+        stringOutput.contains('All tests passed')) {
+      isFirstTestPassed = true;
+    }
+
+    if (isFirstTestPassed &&
+        isReloaded == false &&
+        stringOutput.contains('press "r" to restart')) {
```
Comment (bartekpacia) [inline dev/cli_tests/patrol_develop_test.dart:84]: extract these 3 vars to a single, well-named variable

File: dev/cli_tests/patrol_develop_test.dart [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+      print('Exiting with exit code 0');
+      // TODO: kill `patrol develop` process and its children
+      io.exit(0);
+    }
+
+    inactivityTimer?.cancel();
+
+    if (stringOutput.contains('Completed building')) {
+      inactivityTimer = Timer(const Duration(minutes: 1), () {
+        print('One minute of inactivity, something went wrong...');
```
Comment (bartekpacia) [inline dev/cli_tests/patrol_develop_test.dart:105]: let's replace "one minute" with the value of the `duration` (assign it to a `final`)

File: .cirrus.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
+test_linux_task:
+  name: Test `patrol develop` on Linux
+  container:
+    image: ghcr.io/cirruslabs/flutter:stable
+    cpu: 6
+    memory: 12G
```
Comment (bartekpacia) [inline .cirrus.yml:6]: that's a lot

File: .cirrus.yml [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    PATH: $HOME/.pub-cache/bin:${PATH}
+  timeout_in: 30m
+
+  setup_simulator_script: |
+    xcrun simctl boot "iPhone 14 Pro Max"
+  setup_flutter_script: |
+    flutter channel stable
+    flutter upgrade
+    flutter precache
+    flutter --version
```
Comment (bartekpacia) [inline .cirrus.yml:69]: We'd like to enforce a specific version (version range), e.g. `3.10.x`. You can use `fvm` for this.

File: .cirrus.yml (hunk 9 above)
Comment (fylyppo) [reply .cirrus.yml:6]: with 8 it didn't fit in 10 minutes of test building

Comment (bartekpacia) [reply .cirrus.yml:6]: sounds like a `cpu` bottleneck, not `memory`, but ok

File: .cirrus.yml [hunk 11]
```diff
@@ -65,9 +65,12 @@ test_macos_task:
...
   setup_flutter_script: |
-    flutter channel stable
-    flutter upgrade
-    flutter precache
+    curl -LO https://github.com/fluttertools/fvm/releases/download/2.4.1/fvm-2.4.1-macos-arm64.tar.gz
+    tar -xf fvm-2.4.1-macos-arm64.tar.gz
+    cd fvm
+    ./fvm install 3.7.12
+    ./fvm global 3.7.12
+    ./fvm doctor
```
Comment (bartekpacia) [inline .cirrus.yml:73]: this is macOS - you can simply `brew install fvm`

**Pull request merged**